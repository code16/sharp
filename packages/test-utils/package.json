{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}