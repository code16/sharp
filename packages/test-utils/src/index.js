export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}