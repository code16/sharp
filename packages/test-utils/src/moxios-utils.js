import moxios from 'moxios';

export function nextRequestFulfilled(config, delay) {
    return new Promise(resolve => {
        moxios.wait(async () => {
            let request = moxios.requests.mostRecent();
            let response = await request.respondWith(config);
            resolve(response); // moxios Response object
        }, delay)
    });
}
