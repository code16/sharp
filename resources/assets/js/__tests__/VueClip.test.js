import Vue from 'vue';
import { MockInjections, MockI18n, wait } from "./test-utils";

import VueClip from '../components/form/fields/upload/VueClip';
import SharpModal from '../components/Modal';
import { nextRequestFulfilled } from './test-utils/moxios-utils';

import moxios from 'moxios';


/** Vue clip native methods
// addedFile(File)
// removedFile(File)
// sending()
// complete()
// error()
// uploadProgress()
// thumbnail()
// drop()
// dragEnter()
// dragLeave()
// totalUploadProgress()
// maxFilesExceeded()
// queueComplete()
*/


function mockFile(props) {
    return {
        upload: {},
        ...props,
    }
}

function mockXhrResponse(data) {
    return {
        responseText: JSON.stringify(data)
    }
}

describe('vue-clip',() => {
    Vue.use(MockI18n);

    let toLocaleString = Number.prototype.toLocaleString;
    Number.prototype.toLocaleString = function(locale, options) {
        return toLocaleString.call(this,'en-EN', options)
    };

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-vue-clip :value="value"
                    :options="{ url:'/' }"
                    download-id="my_upload" 
                    pending-key="my_upload.0"
                    :modifiers="{}"
                    :ratio-x="ratioX" 
                    :ratio-y="ratioY"
                    :read-only="readOnly"
                    :croppable-file-types="croppableFileTypes"
                />
            </div>
        `;
        MockI18n.mockLangFunction();
    });


    test('can mount VueClip component', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount read-only VueClip component', async () => {
        await createVm({
            propsData: {
                readOnly: true
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount initially non image VueClip component', async () => {
        await createVm({
            data: () => ({
                value: {
                    name: 'Fichier.pdf',
                    size: 15000
                }
            })
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount initially image VueClip component', async () => {
        await createVm({
            propsData: {
                ratioX: 1, ratioY: 2
            },
            data: () => ({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg',
                    size: 15000
                }
            })
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount non croppable image', async () => {
        await createVm({
            propsData: {
                ratioX: 1, ratioY: 2, croppableFileTypes: ['.png']
            },
            data: () => ({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg',
                    size: 15000
                }
            })
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount "in progress" VueClip component', async () => {
        let $vueClip = await createVm();

        $vueClip.addedFile({
            name: 'Fichier.pdf',
            size: 15000,
            upload: {}
        });

        let { file } = $vueClip;

        file.progress = 50;

        await Vue.nextTick();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount "destroyed" VueClip component', async () => {
        let $vueClip = await createVm();

        $vueClip.$destroy();

        await Vue.nextTick();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('download', async () => {
        let $vueClip = await createVm({
            data: () => ({
                value: {
                    name: 'Fichier.pdf',
                    size: 15000
                }
            })
        });

        moxios.install($vueClip.axiosInstance);

        let { dlLink } = $vueClip.$refs;

        dlLink.onclick = jest.fn();
        URL.createObjectURL = jest.fn(()=>'blob:1234');

        $vueClip.$form.downloadLinkBase = '/sharp/api';

        $vueClip.download();

        let { request } = await nextRequestFulfilled({ status: 200, response: '<<file_content>>' });

        expect(URL.createObjectURL).toHaveBeenCalled();
        expect(dlLink.href).toBe('blob:1234');
        expect(dlLink.onclick).toHaveBeenCalled();

        expect(request.config).toMatchObject({
            method: 'post',
            responseType: 'blob',
            data: JSON.stringify({
                fileName: 'Fichier.pdf'
            }),
            url: '/sharp/api/my_upload'
        });

        moxios.uninstall($vueClip.axiosInstance);
    });

    test('emit update on image loaded', async () => {
        let $vueClip = await createVm({
            data: () => ({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg',
                    size: 15000
                }
            })
        });

        let $image = $vueClip.$el.querySelector('img');
        let handleImageUpdated = jest.fn();

        $vueClip.isNew = true; // simulate new loaded image
        $vueClip.$on('image-updated', handleImageUpdated);

        $image.dispatchEvent(new UIEvent('load'));

        expect(handleImageUpdated).toHaveBeenCalledTimes(1);
    });

    test('init files and reset on value null', async () => {
        let $vueClip = await createVm({
            data: ()=>({
                value: {
                    name: 'Fichier.pdf'
                }
            })
        });

        let { $root:vm } = $vueClip;

        expect($vueClip.files).toMatchObject([{
            name: 'Fichier.pdf'
        }]);

        vm.value = null;

        await Vue.nextTick();

        expect($vueClip.files).toEqual([]);
    });

    test('call function on state changed', async () => {
        let $vueClip = await createVm({
            data: ()=>({
                value: {
                    name: 'Fichier.pdf'
                }
            })
        });

        $vueClip.onStatusAdded = jest.fn();
        $vueClip.onStatusError = jest.fn();
        $vueClip.onStatusSuccess = jest.fn();

        $vueClip.file.status = 'added';
        await Vue.nextTick();
        expect($vueClip.onStatusAdded).toHaveBeenCalledTimes(1);

        $vueClip.file.status = 'error';
        await Vue.nextTick();
        expect($vueClip.onStatusError).toHaveBeenCalledTimes(1);

        $vueClip.file.status = 'success';
        await Vue.nextTick();
        expect($vueClip.onStatusSuccess).toHaveBeenCalledTimes(1);
    });

    test('file property is the first item of "files"', async () => {
        let $vueClip = await createVm({
            data: ()=>({
                value: {
                    name: 'Fichier.pdf'
                }
            })
        });

        expect($vueClip.files[0]).toBe($vueClip.file);
    });

    test('compute images properties correctly', async () => {
        let $vueClip = await createVm({
            data: ()=>({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg'
                }
            })
        });

        expect($vueClip.croppedImg).toBeFalsy();

        expect($vueClip.originalImageSrc).toBe($vueClip.file.thumbnail);
        expect($vueClip.imageSrc).toBe($vueClip.file.thumbnail);

       $vueClip.remove();

       $vueClip.addedFile(mockFile({
           name: 'Fic.pdf',
           dataUrl:'data:image/jpeg,<<source image>>'
       }));

        await Vue.nextTick();

        expect($vueClip.originalImageSrc).toBe($vueClip.file.dataUrl);
        expect($vueClip.imageSrc).toBe($vueClip.file.dataUrl);

        $vueClip.croppedImg = 'data:image/jpeg,<<cropped image>>';

        await Vue.nextTick();

        expect($vueClip.imageSrc).toBe($vueClip.croppedImg);
    });

    test('operations finished without any active', async () =>{
        let $vueClip = await createVm({
            data: ()=>({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg'
                }
            })
        });

        expect($vueClip.operations).toEqual(['crop']);

        expect($vueClip.operationFinished).toEqual({
            crop: null
        });

        expect($vueClip.activeOperationsCount).toBe(0);
    });


    test('operations finished with one active', async () => {
        let $vueClip = await createVm({
            propsData: {
                ratioX: 1,
                ratioY: 1
            },
            data: ()=>({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg'
                }
            })
        });

        expect($vueClip.operationFinished).toEqual({
            crop: false
        });

        expect($vueClip.activeOperationsCount).toBe(1);
        expect($vueClip.operationFinishedCount).toBe(0);


        $vueClip.croppedImg = 'data:image/jpeg';

        expect($vueClip.operationFinishedCount).toBe(1);
        expect($vueClip.operationFinished).toEqual({
            crop: true
        });
    });

    test('progress with operations', async () => {
        let $vueClip = await createVm({
            propsData: {
                ratioX: 1,
                ratioY: 1
            },
            data: ()=>({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg'
                }
            })
        });

        $vueClip.file.progress = 50;

        expect($vueClip.progress).toBe(47.5);

        $vueClip.file.progress = 100;

        expect($vueClip.progress).toBe(95);

        $vueClip.croppedImg = 'data:image/jpeg';

        expect($vueClip.progress).toBe(100);
    });

    test('file size', async () => {
        let $vueClip = await createVm({
            data: ()=>({
                value: {
                    name: 'Fichier.pdf',
                    size: 0
                }
            })
        });

        Number.prototype.toLocaleString = jest.fn(Number.prototype.toLocaleString);

        $vueClip.file.size = 1024*1024/2;

        expect($vueClip.size).toBe('0.5 MB');
        expect(Number.prototype.toLocaleString).toHaveBeenCalled();

        $vueClip.file.size = 1024;
        expect($vueClip.size).toBe('<0.1 MB');
    });

    test('has crop', async () => {
        let $vueClip = await createVm({
            propsData: {
                ratioX:1, ratioY:1
            },
            data: ()=>({
                value: {
                    name: 'Photo2.jpg',
                }
            })
        });

        let { $root:vm } = $vueClip;

        expect($vueClip.hasInitialCrop).toBe(true);
        expect($vueClip.isCroppable).toBe(true);

        vm.ratioX = 0;

        await Vue.nextTick();

        expect($vueClip.hasInitialCrop).toBe(false);

        vm.croppableFileTypes = ['.jpg'];

        await Vue.nextTick();

        expect($vueClip.isCroppable).toBe(true);
        expect($vueClip.hasInitialCrop).toBe(false);

        vm.ratioX = 1;
        await Vue.nextTick();

        expect($vueClip.hasInitialCrop).toBe(true);

        $vueClip.file.name = 'visual.png';

        expect($vueClip.isCroppable).toBe(false);
        expect($vueClip.hasInitialCrop).toBe(false);
    });

    test('file extension', async () => {
        let $vueClip = await createVm({
            data: ()=>({
                value: {
                    name: 'Photo2.jpg'
                }
            })
        });
        expect($vueClip.fileExtension).toEqual('.jpg');
    });

    test('filename', async () => {
        let $vueClip = await createVm({
            data:()=> ({
                value: {
                    name: '/sharp/entities/Image.png'
                }
            })
        });

        expect($vueClip.fileName).toBe('Image.png');

        $vueClip.file.name = 'Photo2.jpg';

        await Vue.nextTick();

        expect($vueClip.fileName).toBe('Photo2.jpg');
    });

    test('set pending', async () => {
        let $vueClip = await createVm({
            data:()=> ({
                value: {
                    name: 'Fichier.pdf'
                }
            })
        });

        let setPendingListener = jest.fn();

        $vueClip.actionsBus.$on('setPendingJob', setPendingListener);

        $vueClip.onStatusAdded();
        expect(setPendingListener).toHaveBeenCalledTimes(1);
        expect(setPendingListener).toHaveBeenCalledWith({
            key: 'my_upload.0',
            origin: 'upload',
            value: true
        });

        $vueClip.file.xhrResponse = mockXhrResponse({});

        $vueClip.onStatusSuccess();
        expect(setPendingListener).toHaveBeenCalledTimes(2);
        expect(setPendingListener).toHaveBeenLastCalledWith({
            key: 'my_upload.0',
            origin: 'upload',
            value: false
        });

        $vueClip.onStatusError();
        expect(setPendingListener).toHaveBeenCalledTimes(3);
        expect(setPendingListener).toHaveBeenLastCalledWith({
            key: 'my_upload.0',
            origin: 'upload',
            value: false
        });
    });

    test('on status added', async () => {
        let $vueClip = await createVm();

        let handleReset = jest.fn();

        $vueClip.$on('reset', handleReset);
        $vueClip.onStatusAdded();

        expect(handleReset).toHaveBeenCalled();
    });

    test('on status error', async () => {
        let $vueClip = await createVm({
            data:()=> ({
                value: {
                    name: 'Fichier.pdf'
                }
            })
        });

        let handleError = jest.fn();
        $vueClip.remove = jest.fn();

        $vueClip.file.errorMessage = "Can't upload";

        $vueClip.$on('error', handleError);
        $vueClip.onStatusError();

        expect(handleError).toHaveBeenCalledTimes(1);
        expect(handleError).toHaveBeenCalledWith("Can't upload");
        expect($vueClip.remove).toHaveBeenCalled();
    });

    test('on status success', async () => {
        let $vueClip = await createVm({
            data:()=> ({
                value: {
                    name: 'Fichier.pdf'
                }
            })
        });

        $vueClip.file.xhrResponse = mockXhrResponse({
            fileName: 'storage/Fichier.pdf'
        });

        let handleSuccess = jest.fn(), handleInput = jest.fn();

        $vueClip.$on('success', handleSuccess);
        $vueClip.$on('input', handleInput);

        $vueClip.onStatusSuccess();

        expect(handleSuccess).toHaveBeenCalledTimes(1);
        expect(handleSuccess).toHaveBeenCalledWith({
            fileName: 'storage/Fichier.pdf',
            uploaded: true
        });

        expect(handleInput).toHaveBeenCalledTimes(1);
        expect(handleInput).toHaveBeenCalledWith({
            fileName: 'storage/Fichier.pdf',
            uploaded: true
        });
    });

    test('remove', async () => {
        let $vueClip = await createVm({
            data:()=> ({
                value: {
                    name: 'Fichier.pdf'
                }
            })
        });

        let handleInput=jest.fn(), handleReset=jest.fn(), handleRemoved=jest.fn();
        $vueClip.$on('input', handleInput);
        $vueClip.$on('reset', handleReset);
        $vueClip.$on('removed', handleRemoved);

        $vueClip.remove();

        expect(handleInput).toHaveBeenCalledTimes(1);
        expect(handleInput).toHaveBeenCalledWith(null);

        expect(handleReset).toHaveBeenCalledTimes(1);
        expect(handleRemoved).toHaveBeenCalledTimes(1);

        expect($vueClip.files).toEqual([]);
        expect($vueClip.canDownload).toBe(false);
    });

    function mockCropper() {
        return {
            cropper: {
                ready: false,
            },
            getCroppedCanvas:jest.fn(()=>({
                toDataURL:jest.fn(()=>'data:image/jpeg')
            })),
            getData:jest.fn(()=>({
                width:100, height:200, x:20, y:40, rotate: 0
            })),
            getImageData:jest.fn(()=>({
                naturalWidth: 200, naturalHeight: 800
            }))
        }
    }

    test('crop', async () => {
        let $vueClip = await createVm({
            propsData: {
                ratioX: 1,
                ratioY: 1
            },
            data:()=> ({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg'
                }
            })
        });

        expect($vueClip.$refs.cropper).toBeTruthy();

        $vueClip.$refs.cropper = mockCropper();
        let { cropper } = $vueClip.$refs.cropper;

        expect($vueClip.isCropperReady()).toBe(false);
        cropper.ready = true;
        expect($vueClip.isCropperReady()).toBe(true);


        $vueClip.file.xhrResponse = mockXhrResponse({});
        $vueClip.onCropperReady = jest.fn($vueClip.onCropperReady);

        $vueClip.onStatusSuccess();

        let handleInput = jest.fn(), handleUpdated = jest.fn();
        $vueClip.$on('input', handleInput);
        $vueClip.$on('updated', handleUpdated);

        await Vue.nextTick();

        expect($vueClip.allowCrop).toBe(true);
        expect($vueClip.isCroppable).toBe(true);
        expect($vueClip.onCropperReady).toHaveBeenCalled();

        expect($vueClip.croppedImg).toBe('data:image/jpeg');

        let testEmit = handler => {
            expect(handler).toHaveBeenCalledTimes(1);
            expect(handler.mock.calls[0][0]).toEqual({
                name: 'Image.jpg',
                thumbnail: '/image.jpg',
                cropData: {
                    width: .5,
                    height: .25,
                    x: .1,
                    y: .05,
                    rotate: -0
                }
            })
        };

        testEmit(handleInput);
        testEmit(handleUpdated);

        $vueClip.$refs.cropper.getData = jest.fn(()=>({
            width: 100, height: 200, x:20, y:40, rotate: -90
        }));

        $vueClip.updateCropData();

        testEmit = handler => {
            expect(handler).toHaveBeenCalledTimes(2);
            expect(handler.mock.calls[1][0]).toMatchObject({
                cropData: {
                    width: 0.125,
                    height: 1,
                    x: 0.025,
                    y: 0.2,
                    rotate: 90
                }
            });
        };

        testEmit(handleInput);
        testEmit(handleUpdated);

        $vueClip.remove();

        expect($vueClip.croppedImg).toBe(null);
    });

    test('cropper modal', async () => {
        let $vueClip = await createVm({
            propsData: {
                ratioX: 1, ratioY:1
            },
            data:()=> ({
                value: {
                    name: 'Image.jpg',
                    thumbnail: '/image.jpg'
                }
            })
        });

        let { modal } = $vueClip.$refs;

        $vueClip.updateCroppedImage = jest.fn();
        $vueClip.updateCropData = jest.fn();

        modal.$emit('ok');

        expect($vueClip.updateCroppedImage).toHaveBeenCalledTimes(1);
        expect($vueClip.updateCropData).toHaveBeenCalledTimes(1);

        let handleInactive = jest.fn(), handleActive = jest.fn();

        $vueClip.$on('inactive', handleInactive);
        $vueClip.$on('active', handleActive);

        $vueClip.onEditButtonClick();

        expect(handleActive).toHaveBeenCalledTimes(1);
        expect($vueClip.showEditModal).toBe(true);


        modal.$emit('hidden');

        expect(handleInactive).toHaveBeenCalledTimes(1);
    });
});

async function createVm(customOptions={}) {
    const vm = new Vue({
        el: '#app',
        mixins: [customOptions, MockInjections],

        components: {
            'sharp-vue-clip': {
                'extends': VueClip,
                components: {
                    'vue-cropper': { render() {} }
                }
            }
        },

        props:['readOnly', 'ratioX', 'ratioY', 'croppableFileTypes'],

        'extends': {
            data:() => ({
                value: null
            })
        }
    });

    await Vue.nextTick();

    return vm.$children[0];
}