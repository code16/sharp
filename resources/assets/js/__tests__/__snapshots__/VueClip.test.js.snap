// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue-clip can mount "destroyed" VueClip component 1`] = `
<div id="app">
	<div class="SharpUpload SharpUpload--empty">
		<div class="SharpUpload__inner">
			<div class="SharpUpload__content">
				<form class="dropzone">
					<button type="button" class="dz-message SharpButton SharpButton--ghost SharpUpload__upload-button">
						{{ form.upload.browse_button }}
					</button>
				</form>
				<!---->
				<div class="clip-preview-template" style="display: none;">
					<div></div>
				</div>
			</div>
		</div>
		<!---->
		<a style="display: none;"></a>
	</div>
</div>
`;

exports[`vue-clip can mount "in progress" VueClip component 1`] = `
<div id="app">
	<div class="SharpUpload">
		<div class="SharpUpload__inner dz-clickable">
			<div class="SharpUpload__content">
				<form class="dropzone" style="display: none;">
					<button type="button" class="dz-message SharpButton SharpButton--ghost SharpUpload__upload-button">
						{{ form.upload.browse_button }}
					</button>
				</form>
				<div class="SharpUpload__container">
					<!---->
					<div class="SharpUpload__infos">
						<div class="mb-3 text-truncate">
							<label class="SharpUpload__filename">Fichier.pdf</label>
							<div class="SharpUpload__info mt-2">
								<span class="mr-2">&lt;0.1 MB</span>
								<a href="" class="SharpUpload__download-link" style="display: none;">
									{{ form.upload.download_link }}
								</a>
							</div>
							<div class="SharpUpload__progress mt-2">
								<div role="progressbar" aria-valuenow="50" aria-valuemin="0" aria-valuemax="100" class="SharpUpload__progress-bar" style="width: 50%;"></div>
							</div>
						</div>
						<div>
							<button type="button" class="SharpButton SharpButton--sm SharpButton--secondary" style="display: none;">
								{{ form.upload.edit_button }}
							</button>
							<button type="button" class="SharpButton SharpButton--sm SharpButton--secondary SharpButton--danger SharpUpload__remove-button">
								{{ form.upload.remove_button }}
							</button>
						</div>
					</div>
					<button type="button" class="SharpUpload__close-button">
						<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpUpload__close-icon">
							<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
						</svg>
					</button>
				</div>
				<div class="clip-preview-template" style="display: none;">
					<div></div>
				</div>
			</div>
		</div>
		<!---->
		<a style="display: none;"></a>
	</div>
</div>
<input type="file" class="dz-hidden-input" style="visibility: hidden; position: absolute; top: 0px; left: 0px; height: 0px; width: 0px;">
`;

exports[`vue-clip can mount VueClip component 1`] = `
<div id="app">
	<div class="SharpUpload SharpUpload--empty">
		<div class="SharpUpload__inner dz-clickable">
			<div class="SharpUpload__content">
				<form class="dropzone">
					<button type="button" class="dz-message SharpButton SharpButton--ghost SharpUpload__upload-button">
						{{ form.upload.browse_button }}
					</button>
				</form>
				<!---->
				<div class="clip-preview-template" style="display: none;">
					<div></div>
				</div>
			</div>
		</div>
		<!---->
		<a style="display: none;"></a>
	</div>
</div>
<input type="file" class="dz-hidden-input" style="visibility: hidden; position: absolute; top: 0px; left: 0px; height: 0px; width: 0px;">
`;

exports[`vue-clip can mount initially image VueClip component 1`] = `
<div id="app">
	<div class="SharpUpload">
		<div class="SharpUpload__inner dz-clickable">
			<div class="SharpUpload__content">
				<form class="dropzone" style="display: none;">
					<button type="button" class="dz-message SharpButton SharpButton--ghost SharpUpload__upload-button">
						{{ form.upload.browse_button }}
					</button>
				</form>
				<div class="SharpUpload__container row">
					<div class="SharpUpload__thumbnail col-4 col-md-4">
						<img src="/image.jpg">
					</div>
					<div class="SharpUpload__infos col-8 col-md-8">
						<div class="mb-3 text-truncate">
							<label class="SharpUpload__filename">Image.jpg</label>
							<div class="SharpUpload__info mt-2">
								<span class="mr-2">&lt;0.1 MB</span>
								<a href="" class="SharpUpload__download-link">
									{{ form.upload.download_link }}
								</a>
							</div>
							<div class="SharpUpload__progress mt-2" style="display: none;">
								<div role="progressbar" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100" class="SharpUpload__progress-bar" style="width: 0%;"></div>
							</div>
						</div>
						<div>
							<button type="button" class="SharpButton SharpButton--sm SharpButton--secondary">
								{{ form.upload.edit_button }}
							</button>
							<button type="button" class="SharpButton SharpButton--sm SharpButton--secondary SharpButton--danger SharpUpload__remove-button">
								{{ form.upload.remove_button }}
							</button>
						</div>
					</div>
					<button type="button" class="SharpUpload__close-button">
						<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpUpload__close-icon">
							<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
						</svg>
					</button>
				</div>
				<div class="clip-preview-template" style="display: none;">
					<div></div>
				</div>
			</div>
		</div>
		<div id="mocked___BV_modal_outer_" style="position: absolute; z-index: 1040;">
			<div id="mocked" role="dialog" aria-hidden="true" aria-labelledby="mocked___BV_modal_title_" aria-describedby="mocked___BV_modal_body_" class="modal fade SharpModal" style="display: none;">
				<div class="modal-dialog modal-md">
					<!---->
					<div role="document" id="mocked___BV_modal_content_" tabindex="-1" class="modal-content">
						<header id="mocked___BV_modal_header_" class="modal-header">
							<div>
								<h5 class="SharpModal__heading">{{ modals.cropper.title }}</h5>
								<button type="button" class="SharpModal__close">
									<svg width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpModal__close-icon">
										<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
									</svg>
								</button>
							</div>
						</header>
						<div id="mocked___BV_modal_body_" class="modal-body">
							<!---->
							<div>
								<button class="SharpButton SharpButton--primary">
									<i class="fa fa-rotate-left"></i>
								</button>
								<button class="SharpButton SharpButton--primary">
									<i class="fa fa-rotate-right"></i>
								</button>
							</div>
						</div>
						<footer id="mocked___BV_modal_footer_" class="modal-footer">
							<button type="button" class="btn btn-secondary">{{ modals.cancel_button }}</button>
							<button type="button" class="btn btn-primary">{{ modals.ok_button }}</button>
						</footer>
					</div>
					<!---->
				</div>
			</div>
			<!---->
		</div>
		<a style="display: none;"></a>
	</div>
</div>
<input type="file" class="dz-hidden-input" style="visibility: hidden; position: absolute; top: 0px; left: 0px; height: 0px; width: 0px;">
`;

exports[`vue-clip can mount initially non image VueClip component 1`] = `
<div id="app">
	<div class="SharpUpload">
		<div class="SharpUpload__inner dz-clickable">
			<div class="SharpUpload__content">
				<form class="dropzone" style="display: none;">
					<button type="button" class="dz-message SharpButton SharpButton--ghost SharpUpload__upload-button">
						{{ form.upload.browse_button }}
					</button>
				</form>
				<div class="SharpUpload__container">
					<!---->
					<div class="SharpUpload__infos">
						<div class="mb-3 text-truncate">
							<label class="SharpUpload__filename">Fichier.pdf</label>
							<div class="SharpUpload__info mt-2">
								<span class="mr-2">&lt;0.1 MB</span>
								<a href="" class="SharpUpload__download-link">
									{{ form.upload.download_link }}
								</a>
							</div>
							<div class="SharpUpload__progress mt-2" style="display: none;">
								<div role="progressbar" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100" class="SharpUpload__progress-bar" style="width: 0%;"></div>
							</div>
						</div>
						<div>
							<button type="button" class="SharpButton SharpButton--sm SharpButton--secondary" style="display: none;">
								{{ form.upload.edit_button }}
							</button>
							<button type="button" class="SharpButton SharpButton--sm SharpButton--secondary SharpButton--danger SharpUpload__remove-button">
								{{ form.upload.remove_button }}
							</button>
						</div>
					</div>
					<button type="button" class="SharpUpload__close-button">
						<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpUpload__close-icon">
							<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
						</svg>
					</button>
				</div>
				<div class="clip-preview-template" style="display: none;">
					<div></div>
				</div>
			</div>
		</div>
		<!---->
		<a style="display: none;"></a>
	</div>
</div>
<input type="file" class="dz-hidden-input" style="visibility: hidden; position: absolute; top: 0px; left: 0px; height: 0px; width: 0px;">
`;

exports[`vue-clip can mount non croppable image 1`] = `
<div id="app">
	<div class="SharpUpload">
		<div class="SharpUpload__inner dz-clickable">
			<div class="SharpUpload__content">
				<form class="dropzone" style="display: none;">
					<button type="button" class="dz-message SharpButton SharpButton--ghost SharpUpload__upload-button">
						{{ form.upload.browse_button }}
					</button>
				</form>
				<div class="SharpUpload__container row">
					<div class="SharpUpload__thumbnail col-4 col-md-4">
						<img src="/image.jpg">
					</div>
					<div class="SharpUpload__infos col-8 col-md-8">
						<div class="mb-3 text-truncate">
							<label class="SharpUpload__filename">Image.jpg</label>
							<div class="SharpUpload__info mt-2">
								<span class="mr-2">&lt;0.1 MB</span>
								<a href="" class="SharpUpload__download-link">
									{{ form.upload.download_link }}
								</a>
							</div>
							<div class="SharpUpload__progress mt-2" style="display: none;">
								<div role="progressbar" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100" class="SharpUpload__progress-bar" style="width: 0%;"></div>
							</div>
						</div>
						<div>
							<button type="button" disabled="disabled" class="SharpButton SharpButton--sm SharpButton--secondary">
								{{ form.upload.edit_button }}
							</button>
							<button type="button" class="SharpButton SharpButton--sm SharpButton--secondary SharpButton--danger SharpUpload__remove-button">
								{{ form.upload.remove_button }}
							</button>
						</div>
					</div>
					<button type="button" class="SharpUpload__close-button">
						<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpUpload__close-icon">
							<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
						</svg>
					</button>
				</div>
				<div class="clip-preview-template" style="display: none;">
					<div></div>
				</div>
			</div>
		</div>
		<!---->
		<a style="display: none;"></a>
	</div>
</div>
<input type="file" class="dz-hidden-input" style="visibility: hidden; position: absolute; top: 0px; left: 0px; height: 0px; width: 0px;">
`;

exports[`vue-clip can mount read-only VueClip component 1`] = `
<div id="app">
	<div class="SharpUpload SharpUpload--empty SharpUpload--disabled">
		<div class="SharpUpload__inner dz-clickable">
			<div class="SharpUpload__content">
				<form class="dropzone">
					<button type="button" disabled="disabled" class="dz-message SharpButton SharpButton--ghost SharpUpload__upload-button">
						{{ form.upload.browse_button }}
					</button>
				</form>
				<!---->
				<div class="clip-preview-template" style="display: none;">
					<div></div>
				</div>
			</div>
		</div>
		<!---->
		<a style="display: none;"></a>
	</div>
</div>
<input type="file" class="dz-hidden-input" style="visibility: hidden; position: absolute; top: 0px; left: 0px; height: 0px; width: 0px;">
`;
